File: dotnet/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example4/Singleton.cs
namespace SingletonDesignPattern.Example4;

public class Singleton
{
	private static Singleton? _instance;
	private static readonly object _lock = new object();
	public string Value { get; set; } = string.Empty;

	private Singleton() { }

	public static Singleton GetInstance(string value)
	{
		if(_instance == null)
		{
			lock(_lock)
			{
				if(_instance == null)
				{
					_instance = new Singleton();
					_instance.Value = value;
				}
			}
		}
		return _instance;
	}

}

File: nodejs/DesignPatterns/CreationalDesignPatterns/SingletonDesignPattern/SingletonDesignPattern.Example4/src/Singleton.ts
import { Mutex } from 'async-mutex';

// Double-checked locking pattern with a value parameter.
export class Singleton {
  private static _instance: Singleton | null = null;
  private static readonly _mutex = new Mutex();
  public value: string = '';

  private constructor() {}

  static async getInstance(value: string): Promise<Singleton> {
    if (!Singleton._instance) {
      const release = await Singleton._mutex.acquire();
      try {
        if (!Singleton._instance) {
          Singleton._instance = new Singleton();
          Singleton._instance.value = value;
        }
      } finally {
        release();
      }
    }
    return Singleton._instance;
  }
}
